lib/auth/auth.go

```diff
@@ -1137,6 +1143,117 @@ func (a *Server) GetMFAAuthenticateChallenge(user string, password []byte) (*MFA
 	return chal, nil
 }
 
+// CreateAuthenticationChallenge creates and returns challenges for a users MFA devices.
+func (a *Server) CreateAuthenticationChallenge(ctx context.Context, req *proto.CreateAuthenticationChallengeRequest) (*proto.MFAAuthenticateChallenge, error) {
+	username := ""
+
+	// Determine how to verify user.
+	switch req.GetRequest().(type) {
+	case *proto.CreateAuthenticationChallengeRequest_UserPwd:
+		username = req.GetUserPwd().Username
+		err := a.WithUserLock(username, func() error {
+			return a.checkPasswordWOToken(username, req.GetUserPwd().Password)
+		})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+	case *proto.CreateAuthenticationChallengeRequest_TokenID:
+		token, err := a.getRecoveryStartToken(ctx, req.GetTokenID())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		username = token.GetUser()
+
+	default:
+		return nil, trace.BadParameter("missing request parameter")
+	}
+
+	protoChal, err := a.mfaAuthChallenge(ctx, username, a.Identity)
+	return protoChal, trace.Wrap(err)
+}
+
+// GetMFADevices returns all mfa devices for the user defined in the token.
+func (a *Server) GetMFADevices(ctx context.Context, req *proto.GetMFADevicesRequest) (*proto.GetMFADevicesResponse, error) {
+	username := ""
+
+	// Determine how to verify user.
+	switch req.GetRequest().(type) {
+	case *proto.GetMFADevicesRequest_TokenID:
+		token, err := a.GetUserToken(ctx, req.GetTokenID())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Restrict to certain kinds of user token and usage.
+		if !isTokenRecoveryApprovedFor2FA(token) {
+			return nil, trace.BadParameter("invalid token")
+		}
+
+		if token.Expiry().Before(a.clock.Now().UTC()) {
+			return nil, trace.BadParameter("expired token")
+		}
```

Comment:
Suggestion: push this logic into the token class, or refactor to an utility method?

Eg:

```go
err := token.Verify(allowedUsages, currentTime)
// verify err
```